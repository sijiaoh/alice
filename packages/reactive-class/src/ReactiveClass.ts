export interface Store {
  onChange: () => void;
}

export interface ReactiveClassConstructorParams {
  store: Store;
}

export class ReactiveClass {
  constructor({ store }: ReactiveClassConstructorParams) {
    this.store = store;
    return new Proxy(this, {
      set: (target, p: keyof ReactiveClass, value: never) => {
        // eslint-disable-next-line no-param-reassign
        target[p] = value;
        this.store.onChange();
        return true;
      },
    });
  }

  private store: Store;
}
