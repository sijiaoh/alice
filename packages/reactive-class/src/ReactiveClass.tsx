/* eslint-disable react-hooks/rules-of-hooks */

import autobind from 'autobind-decorator';
import { produce } from 'immer';
import React, { useEffect, useState } from 'react';
import { Props } from 'react-utils';
import {
  atom,
  RecoilState,
  useRecoilCallback,
  RecoilValue,
  Loadable,
  useRecoilState,
  RecoilRoot,
  useRecoilTransactionObserver_UNSTABLE,
} from 'recoil';

@autobind
export class ReactiveClass<T> {
  static readonly notProvideError = 'ReactiveClass.Provider not provided.';
  static get?: <U>(recoilValue: RecoilValue<U>) => Loadable<U>;
  static set?: <U>(
    recoilVal: RecoilState<U>,
    valOrUpdater: U | ((currVal: U) => U)
  ) => void;

  static Component({ children }: Props) {
    const dummy = useRecoilCallback(({ snapshot, set }) => {
      ReactiveClass.get = snapshot.getLoadable;
      ReactiveClass.set = set;
      return () => {};
    });

    useRecoilTransactionObserver_UNSTABLE(({ snapshot }) => {
      ReactiveClass.get = snapshot.getLoadable;
    });

    const [initialized, setInitialized] = useState(false);
    useEffect(() => {
      // jest用呼び出し。
      dummy();
      setInitialized(true);
    }, [dummy]);

    return <>{initialized ? children : null}</>;
  }

  static Provider({ children }: Props) {
    const { Component } = ReactiveClass;
    return (
      <RecoilRoot>
        <Component>{children}</Component>
      </RecoilRoot>
    );
  }

  readonly state: RecoilState<T>;

  get data() {
    return this.getData();
  }
  set data(data: Readonly<T>) {
    this.setData(data);
  }

  constructor(key: string, defaultStore: Readonly<T>) {
    this.state = atom({ key, default: defaultStore });
  }

  getData() {
    if (!ReactiveClass.get) throw new Error(ReactiveClass.notProvideError);
    return ReactiveClass.get(this.state).getValue();
  }

  setData(state: Readonly<T>) {
    if (!ReactiveClass.set) throw new Error(ReactiveClass.notProvideError);
    ReactiveClass.set(this.state, state);
  }

  changeValue(callback: (state: T) => Readonly<T> | void) {
    this.setData(produce(this.getData(), callback));
  }

  useState() {
    return useRecoilState(this.state);
  }
}
