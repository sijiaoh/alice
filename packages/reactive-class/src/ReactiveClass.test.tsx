import { ReactiveClass } from './ReactiveClass';

class SomeClass extends ReactiveClass<{
  num: number;
  str: string;
  obj: { key: number };
}> {
  constructor() {
    super({ num: 0, str: '0', obj: { key: 0 } });
  }
}

describe(ReactiveClass.name, () => {
  describe(ReactiveClass.prototype.changeData.name, () => {
    it('will update data', () => {
      const some = new SomeClass();

      some.changeData((data) => {
        data.num += 1;
        data.str = '1';
        data.obj.key = 1;
      });

      expect(some.data.num).toBe(1);
      expect(some.data.str).toBe('1');
      expect(some.data.obj.key).toBe(1);

      some.data = {
        ...some.data,
        num: some.data.num + 1,
        str: '2',
      };

      expect(some.data.num).toBe(2);
      expect(some.data.str).toBe('2');
      expect(some.data.obj.key).toBe(1);
    });
  });
});
