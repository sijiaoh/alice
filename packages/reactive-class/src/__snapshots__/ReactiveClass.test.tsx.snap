// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactiveClass  rerender component when selected value changed 1`] = `
Array [
  "0",
  "0",
]
`;

exports[`ReactiveClass  rerender component when selected value changed 2`] = `
Array [
  "hello",
  "0",
]
`;

exports[`ReactiveClass  rerender component when selected value changed 3`] = `
Array [
  "hello",
  "0",
]
`;

exports[`ReactiveClass  rerender component when selected value changed 4`] = `
Array [
  "hello2",
  "1",
]
`;
